src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -93,41 +93,77 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
+        public async ValueTask<ImmutableArray<PackageSource>?> TryGetPackageSourcesAsync(bool allowSwitchToMainThread, CancellationToken cancellationToken)
         {
-            // Only read from _packageSources once, since OnSourceProviderSourcesChanged could reset it to default at
-            // any time while this method is running.
-            var packageSources = _packageSources;
-            if (packageSources != null)
+            // Only read from _packageSourcesAsync once, since OnSourceProviderSourcesChanged could reset it to default
+            // at any time while this method is running.
+            JoinableTask<ImmutableArray<PackageSource>> packageSourcesAsync;
+            lock (_gate)
             {
-                return packageSources;
+                if (_packageSourcesAsync is null)
+                {
+                    _packageSourcesAsync = ThreadingContext.JoinableTaskFactory.RunAsync(() => GetPackageSourcesImplAsync());
+                }
+
+                packageSourcesAsync = _packageSourcesAsync;
             }
 
-            try
+            if (packageSourcesAsync.IsCompleted)
             {
-                packageSources = _packageSourceProvider.Value.GetSources(includeUnOfficial: true, includeDisabled: false)
-                    .SelectAsArray(r => new PackageSource(r.Key, r.Value));
+                // Since the task is already completed, we know this 'await' will complete synchronously.
+                return await packageSourcesAsync;
             }
-            catch (Exception ex) when (ex is InvalidDataException || ex is InvalidOperationException)
+            else if (allowSwitchToMainThread)
             {
-                // These exceptions can happen when the nuget.config file is broken.
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                return await packageSourcesAsync.JoinAsync(cancellationToken).ConfigureAwait(false);
             }
-            catch (ArgumentException ae) when (FatalError.ReportWithoutCrash(ae))
+            else
             {
-                // This exception can happen when the nuget.config file is broken, e.g. invalid credentials.
-                // https://github.com/dotnet/roslyn/issues/40857
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                // The result was not available and switching to the main thread is not allowed. Return without caching
+                // a result.
+                return null;
             }
+        }
 
-            var previousPackageSources = ImmutableInterlocked.InterlockedCompareExchange(ref _packageSources, packageSources, default);
-            if (previousPackageSources != null)
+        private async Task<ImmutableArray<PackageSource>> GetPackageSourcesImplAsync()
+        {
+            // Automatically retry the operation if it gets cancelled due to a change in the workspace.
+            while (true)
```

Comment:
why do we need this?  if the workspace changed, wouldn't we simply compute the package sources the next time we get called?

Reply:
We don't have anything to return if the operation fails for this cancellation.

* Empty isn't valid, because it looks like a successful result even though it didn't complete.
* Cancellation exception isn't valid because the caller did not request it.

This retry loop is the easiest way to ensure the operation completes in a reasonable state.

Reply:
> Empty isn't valid, because it looks like a successful result even though it didn't complete.

Empty would be totally valid for me.  Note: we only cancel when a config change happens.  And in that case, we drop this task anyways.  So new callers will come in and see hte new task with the new values.  Any existing callers will see an empty array, which is 100% fine.